### ISSUE TITLE
Sending `<down>` or `<up>` in Firefox's address bar doesn't work

### ISSUE THREAD

**Issue by SOberhoff** [autokey/autokey#177]
I just installed AutoKey for the first time and rebound a couple of keys via phrases. In particular, I made it so that `alt+i/j/k/l` maps to the arrow keys. I can now move the cursor freely left and right in the address bar. But if I attempt to move down to go through the suggestions, I just see a brief flicker and no action. The same also occurs in other places such as Google's search field. I can move up and down in many other contexts though.

**Comment by josephj11** [comment]
Some applications sometimes make our lives difficult - notably Firefox and LibreOffice.

First, just to make sure we're on the same page, please take a look at [our problem reporting guide](https://github.com/autokey/autokey/wiki/Problem-Reporting-Guide).

We need more information to solve this issue.

**Comment by luziferius** [comment]
The flickering you experience is a brief window focus loss. Or in X11-terms, you see a `FocusOut` event immediately followed by a `FocusIn` Event. (Try open a console, start `xev` and experiment with it. It shows all events that happen when you press keys.)

When Firefox sees a `FocusOut`, it closes all overlay windows (like the one below the adress bar you want to navigate…)

I think this focus switch happens, because AutoKey has to remap the keyboard layout. This is required to remove the \<Alt> key from your keyboard for the phrase expansion duration. Otherwise you’d get \<Alt>+\<left>, because you are physically holding \<Alt> on your keyboard.


I don’t know if this is solvable using our [currently used libraries](https://www.x.org/releases/X11R7.6/doc/libXtst/recordlib.html).

**Comment by SOberhoff** [comment]
The problem reporting guide appears empty to me.

On another note: I've made a number of fruitless attempts to remap these keys in the way I've described by directly altering xkb configuration files. Is this issue something I could avoid if I managed to get that to work?

**Comment by josephj11** [comment]
The problem reporting guide is MIA. I'm looking for it.

**Comment by luziferius** [comment]
Hmm. Can you try, if this still happens with version 0.95.4? You can grab debian packages from the github release page or install using pip3 from a local checkout.
Maybe it is related to #181, which got fixed in the latest release.

**Comment by feluxe** [comment]
I have the same issue using `autokey 95.4`. It worked until some weeks ago. I don't know what caused it to stop working. I guess it came with an upgrade of my linux  system.

The following shortcut doesn't work when I'm in the address bar:

```python
keyboard.send_keys("<down>")
```
hotkey: `<ctrl>+n`
window filter: `.*irefox.*`

Strangely, when I change the shortcut to:

```python
keyboard.send_keys("a")
```
it works. It writes `a` into the address bar when I hit `ctrl+n`.

**Comment by AdrianSkar** [comment]
Same problem here; seems to happen only on specific apps like Firefox navbar, webpages (like Google sheets) however seem to get arrows properly.
I don't know if this is an expected behavior but [I have CapsLock trigggering arrows](https://github.com/AdrianSkar/keybindings/tree/master/linux/autokey/ijkl) on `ijkl` and Home/End on `uo` and the weird thing is that **xev** logs _Home_  and _Hyper_L_ as being pressed when using `CapsLock + u` but doesn't log _Down_ when using `CapsLock + k`; it logs _k_ and _Hyper_L_  instead.

**Comment by josephj11** [comment]
@AdrianSkar  Aside from the rest of this bug: If you have repurposed/remapped CapsLock, you should select Settings->General->Modifier keys->Disable handling of Capslock key.

Also,  most unmodified keyboard layouts do not have a Hyper key defined. If you have not modified your keyboard layout, seeing Hyper is pretty weird.

If you are intentionally using the Hyper key for something, Dell Anderson created a short video about that which is linked in our Wiki under Advanced Scripting.

**Comment by AdrianSkar** [comment]
Hi @josephj11!
> @AdrianSkar Aside from the rest of this bug: If you have repurposed/remapped CapsLock, you should select Settings->General->Modifier keys->Disable handling of Capslock key.

I'm not sure what this option does but it doesn't change the behavior (nor the xev output) I get whether I enable/disable it (I have _Make Caps Lock and addditional Hyper_ in KDE in order to use these shortcuts).
> Also, most unmodified keyboard layouts do not have a Hyper key defined. If you have not modified your keyboard layout, seeing Hyper is pretty weird.

I have it because its the one I've seen works well from other related remapping tutorials/info I've found + the option was already available in KDE (like making it behave like an additional Super). Should I expect it to behave better/differently changing it to a different one?

> If you are intentionally using the Hyper key for something, Dell Anderson created a short video about that which is linked in our Wiki under Advanced Scripting.

I'll try using xmodmap like that to see if it works better but, if I'm not mistaken, this wouldn't solve this particular issue, right?

Thank you!

**Comment by josephj11** [comment]
@AdrianSkar AutoKey tends to not recognize a repurposed Capslock key, so you should use that  option even if it does not affect this issue.

In general, AutoKey does not like remapped keyboards. It seems to use the raw keypresses and ignores the remapping sometimes. I'm not sure of the exact details, but we have seen similar issues a number of times.

Xmodmap is just a different (very nice) utility for remapping your keyboard, so I wouldn't expect it to help.

If you want to, you can experiment with using xdotool or the Python xdo module by calling it from your script. It uses different libraries than AutoKey does and sometimes works when the AutoKey API won't.

Also, Firefox's address bar is a favorite place for things to go wrong. If you are doing more than one keyboard event at a time, you should add time.sleep(0.5) calls between them. If that helps, then you can reduce the delays until it breaks again so you will know how much delay is actually needed, but don't be too precise because the required delays may change depending upon the current load on your system.

**Comment by AdrianSkar** [comment]
@josephj11 I'll definetly experiment with that. Thank you!

**Comment by RolandSherwin** [comment]
Having the same issue, but it works without any problem when I use AnyDesk to remotely access my pc and then use `alt+k` to send the down arrow key.

**Comment by sebastiansam55** [comment]
interesting, you might want to try adding a bit of a delay to your script
to allow time for you to release ctrl? I think that may have come up before
somewhere

**Comment by RolandSherwin** [comment]
I have tried adding `time.sleep()` with values ranging from `1` to `0.001` with both `send_key()` and `send_keys()` but with no luck.

**Comment by barsv** [comment]
TLDR: I solved it for myself but without autokey. I tried several options with different drawbacks. I'm using now the option#3.

Long story:

### option#1 - kbct
I was able to made Ctrl+Alt+hjkl to work using https://github.com/samvel1024/kbct
It was not very easy though. Installation instructions are not complete for ubuntu. I had to use the information from this pull request: https://github.com/samvel1024/kbct/pull/20/files to create `/etc/systemd/system/kbct.service` file and then manually start the service with `sudo systemctl start kbct`. My config: https://gist.github.com/barsv/677aa1807eec57d2d934d227c9418652
I had to map Ctl+Alt+hjkl to arrows instead of just Alt+hjkl because kbct sends Alt release event to Firefox and it triggers the top menu.
The drawback of this solution is that I can't use Ctrl+l to move to the next word.
I tried to remap right alt to something not used like fn24 but fn24 seems to not work as a modifier.
If you are fine to not be able to use Ctrl with movements using hjkl then this solution might be good.

### option#2 - xmodmap
I was able to remap CapsLock+hjkl to arrows and map Shift+CapsLock to CapsLock by following https://superuser.com/a/1288390 
```
clear lock
clear mod5
add lock = Mode_switch
keycode 66 = Mode_switch Caps_Lock NoSymbol NoSymbol
keycode 43 = h H Left Left
keycode 44 = j J Down Down
keycode 45 = k K Up Up
keycode 46 = l L Right Right
```
This solution is not good for me because it breaks my second layout for Russian. When I switch to Russian I can't use hjkl keys. Also I have a LED indicator on my CapsLock and it goes out of sync with the actual status of CapsLock when I use this approach.
CapsLock LED issue is not a problem if I use right Alt instead of CapsLock:
```
clear mod1
clear mod5
keycode 108 = Mode_switch
add mod1 = Alt_L
add mod5 = Mode_switch
keycode 43 = h H Left Left
keycode 44 = j J Down Down
keycode 45 = k K Up Up
keycode 46 = l L Right Right
```
And in theory I think that it should be possible to use 5th and 6th keysyms like this:
```
clear mod1
clear mod5
keycode 108 = ISO_Level3_Shift
add mod1 = Alt_L
add mod5 = ISO_Level3_Shift
keycode  42 = g G Cyrillic_pe Cyrillic_PE Home Home
keycode  43 = h H Cyrillic_er Cyrillic_ER Left Left
keycode  44 = j J Cyrillic_o Cyrillic_O Down Down
keycode  45 = k K Cyrillic_el Cyrillic_EL Up Up
keycode  46 = l L Cyrillic_de Cyrillic_DE Right Right
```
But for some reason it doesn't work for me. I wasn't able to figure out why.
Maybe it's just not possible https://unix.stackexchange.com/a/313711/479942 but I'm not sure.
So if you use only one layout (language) then this solution might be the best option. It is easy to setup, doesn't need sudo and easy to move to another machine with other dotfiles.

### option#3 - XKB
XKB has complicated rules and I had to follow several tutorials to get some understanding on how to use xkbcomp tool (https://superuser.com/a/1737465/301114)
I still don't understand XKB completely, but I was able to configure what I want. I did the following. First I did a dump of my existing config using `xkbcomp $DISPLAY output.xkb`. Then I added the following at the end of xkb_symbols section:
```
    key <RALT>  { 
        [ NoSymbol, NoSymbol ],
        actions[Group1]=[
            SetControls(controls=overlay1), 
            LockControls(controls=overlay1)
        ],
        actions[Group2]=[
            SetControls(controls=overlay1), 
            LockControls(controls=overlay1)
        ]
    };
    key <AC07> { overlay1=<DOWN> };// j
    key <AC08> { overlay1=<UP> };// k
    key <AC09> { overlay1=<RGHT> };// l
    key <AC06> { overlay1=<LEFT> };// h
    key <AC10> { overlay1=<HOME> };// ;
    key <AC11> { overlay1=<END> };// '
    key <AD10> { overlay1=<PGUP> };// p
    key <AB06> { overlay1=<PGDN> };// n
    key <AB02> { overlay1=<DELE> };// x
    key <AD08> { overlay1=<ESC> };// i
```
Saved it as my.xkb and then passed back to xkbcomp: `xkbcomp my.xkb $DISPLAY`.
So far I haven't found bugs with it. The only disadvantage is that I don't have right Alt(RALT) now. The way this config works is that it creates an overlay for the keyboard. To activate the overlay I can either hold RALT or press Shift+RALT to lock it similarly to how CapsLock locks Shift.
I also added a startup script to apply my config after system boot. I use GNOME so I used `.config/autostart'. Actually I used GUI for this at first but it just created files in `.config/autostart'. Now I have 3 files there: my.xkb, my-xkb.sh, my-xkb.sh.desktop. 
my.xkb is the file with config. my-xkb.sh is the following script:
```
#!/bin/bash
# load my xkb config for Alt+hjkl navigation.
logger "started my-xkb.sh"
{
    # sleep is needed because otherwise GNOME reloads
    # own keyboard settings.
    sleep 5
    logger "starting xkbcomp in my-xkb.sh"
    xkbcomp ~/.config/autostart/my.xkb $DISPLAY
    logger "done xkbcomp in my-xkb.sh"
} &
```
and my-xkb.sh.desktop is a config file for GNOME to run my-xkb.sh script:
```
[Desktop Entry]
Type=Application
Exec=bash -c '$HOME/.config/autostart/my-xkb.sh'
Hidden=false
NoDisplay=false
X-GNOME-Autostart-enabled=true
Name[en_US]=my-xkb
Name=hjkl
Comment[en_US]=Load my custom xkb config for Alt+hjkl navigation.
Comment=
```

**Comment by josephj11** [comment]
@barsv Wow. Thanks for documenting all your extensive work.
